packages/fxa-auth-server/lib/payments/subscription-reminders.ts

```diff
@@ -0,0 +1,216 @@
+  year(plan: Plan, planLength: number): Boolean;
+}
+
+const eligiblePlansByInterval: EligiblePlansByInterval = {
+  day: (plan, planLength) => plan.interval_count >= planLength,
+  week: (plan, planLength) =>
+    plan.interval_count >= planLength / DAYS_IN_A_WEEK,
+  month: (plan, planLength) =>
+    plan.interval_count >= planLength / DAYS_IN_A_MONTH,
+  year: (plan, planLength) =>
+    plan.interval_count >= planLength / DAYS_IN_A_YEAR,
+};
+
+export class SubscriptionReminders {
+  private db: any;
+  private mailer: any;
+  private stripeHelper: StripeHelper;
+
+  constructor(
+    private log: Logger,
+    config: ConfigType,
+    private planLength: number = DEFAULT_PLAN_LENGTH,
+    private reminderLength: number = DEFAULT_REMINDER_LENGTH,
+    db: any,
+    mailer: any
+  ) {
+    this.db = db;
+    this.mailer = mailer;
+    this.stripeHelper = Container.get(StripeHelper);
+  }
+
+  /**
+   * For all possible plan.intervals, determine if the plan is sufficiently
+   * long based on planLength.
+   */
+  private isEligiblePlan(plan: Plan): boolean {
+    for (let [interval, checkFn] of Object.entries(eligiblePlansByInterval)) {
+      if (plan.interval === interval && checkFn(plan, this.planLength)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  private async getEligiblePlans(): Promise<Plan[]> {
+    const allPlans = await this.stripeHelper.allPlans();
+    return allPlans.filter((plan) => this.isEligiblePlan(plan));
+  }
+
+  /**
+   * Returns a window of time in seconds [startTimeS, endTimeS)
+   * that is exactly one day, reminderLength days from now in UTC.
+   */
+  private getStartAndEndTimes(reminderLengthMs: number): TimeSpanInS {
+    const reminderDay = new Date(Date.now() + reminderLengthMs);
+    // Get hour 0, minute 0, second 0 for today's date
+    const startingTimestamp = new Date(
+      Date.UTC(
+        reminderDay.getUTCFullYear(),
+        reminderDay.getUTCMonth(),
+        reminderDay.getUTCDate(),
+        0,
+        0,
+        0
+      )
+    );
+    // Get hour 0, minute, 0, second 0 for one day from today's date
+    const endingTimestamp = new Date(startingTimestamp.getTime() + MS_IN_A_DAY);
+
+    const startTimeS = Math.floor(startingTimestamp.getTime() / 1000);
+    const endTimeS = Math.floor(endingTimestamp.getTime() / 1000);
+    return {
+      startTimeS,
+      endTimeS,
+    };
+  }
+
+  private async alreadySentEmail(
+    uid: string,
+    currentPeriodStartMs: number,
+    emailParams: SentEmailParams
+  ) {
+    const emailRecord = await SentEmail.findLatestSentEmailByType(
+      uid,
+      EMAIL_TYPE,
+      emailParams
+    );
+    // This could be the first email for a given subscription or a subsequent one.
+    return !!emailRecord && emailRecord.sentAt > currentPeriodStartMs;
+  }
+
+  private async updateSentEmail(uid: string, emailParams: SentEmailParams) {
+    await SentEmail.createSentEmail(uid, EMAIL_TYPE, emailParams);
+  }
```

Comment:
We may want to import so it can be used elsewhere. Where would be the best place for this? I don’t see an obvious place in the current fxa-auth-server file structure (e.g. in ./lib).

Reply:
What would be the interface for the exported `updateSentEmail`?  If it's the same, then it'd only work with the 'subscriptionRenewalReminder' email type.  In that case, I hope people are using this class to begin with.

Reply:
True; I guess I was thinking I'd change the params to accept an `emailType`. Probably not worth worrying about now as we only use this in one place so far.